x/protorev/keeper/hooks.go

```diff
@@ -0,0 +1,215 @@
+}
+
+// AfterLastPoolPositionRemoved is a noop.
+func (h Hooks) AfterLastPoolPositionRemoved(ctx sdk.Context, sender sdk.AccAddress, poolId uint64) {
+}
+
+// AfterConcentratedPoolSwap stores swaps to be checked by protorev given the coins swapped in the pool.
+func (h Hooks) AfterConcentratedPoolSwap(ctx sdk.Context, sender sdk.AccAddress, poolId uint64, input sdk.Coins, output sdk.Coins) {
+	// Checked to avoid future unintended behavior based on how the hook is called
+	if len(input) != 1 || len(output) != 1 {
+		return
+	}
+
+	h.k.storeSwap(ctx, poolId, input[0].Denom, output[0].Denom)
+}
+
+// ----------------------------------------------------------------------------
+// HELPER METHODS
+// ----------------------------------------------------------------------------
+
+// afterPoolCreatedWithCoins checks if the new pool should be stored as the highest liquidity pool
+// for any of the base denoms, and stores it if so.
+func (k Keeper) afterPoolCreatedWithCoins(ctx sdk.Context, poolId uint64) {
+	baseDenoms, err := k.GetAllBaseDenoms(ctx)
+	if err != nil {
+		ctx.Logger().Error("Protorev error getting base denoms in AfterCFMMPoolCreated hook", err)
+		return
+	}
+
+	baseDenomMap := make(map[string]bool)
+	for _, baseDenom := range baseDenoms {
+		baseDenomMap[baseDenom.Denom] = true
+	}
+
+	pool, err := k.poolmanagerKeeper.GetPool(ctx, poolId)
+	if err != nil {
+		ctx.Logger().Error("Protorev error getting pool in AfterCFMMPoolCreated hook", err)
+		return
+	}
+
+	denoms, err := k.poolmanagerKeeper.RouteGetPoolDenoms(ctx, poolId)
+	if err != nil {
+		ctx.Logger().Error("Protorev error getting pool liquidity in afterPoolCreated", err)
+		return
+	}
+
+	// Pool must be active and the number of denoms must be 2
+	if pool.IsActive(ctx) && len(denoms) == 2 {
+		if baseDenomMap[denoms[0]] {
+			k.compareAndStorePool(ctx, poolId, denoms[0], denoms[1])
+		}
+		if baseDenomMap[denoms[1]] {
+			k.compareAndStorePool(ctx, poolId, denoms[1], denoms[0])
+		}
+	}
+}
+
+// storeSwap stores a swap to be checked by protorev when attempting backruns.
+func (k Keeper) storeSwap(ctx sdk.Context, poolId uint64, tokenIn, tokenOut string) {
+	swapToBackrun := types.Trade{
+		Pool:     poolId,
+		TokenIn:  tokenIn,
+		TokenOut: tokenOut,
+	}
+
+	if err := k.AddSwapsToSwapsToBackrun(ctx, []types.Trade{swapToBackrun}); err != nil {
+		ctx.Logger().Error("Protorev error adding swap to backrun from storeSwap", err) // Does not return since logging is last thing in the function
+	}
+}
+
+// compareAndStorePool compares the liquidity of the new pool with the liquidity of the stored pool, and stores the new pool if it has more liquidity.
+func (k Keeper) compareAndStorePool(ctx sdk.Context, poolId uint64, baseDenom, otherDenom string) {
+	storedPoolId, err := k.GetPoolForDenomPair(ctx, baseDenom, otherDenom)
+	if err != nil {
+		// Error means no pool exists for this pair, so we set it
+		k.SetPoolForDenomPair(ctx, baseDenom, otherDenom, poolId)
+		return
+	}
```

Comment:
Why do we have `storedPoolId` and `poold` here? What if for the same denom pair we have multiple pools stored - CL and balancer but the hoos is executed for one but not the other?

Reply:
The protorev module currently only stores a single pool for a given denom pair (irrespective of type).

So when we error here that means no pool is stored for the pair (irrespective of type), and then we proceed to store the poolId of the pool that was just created and triggered the hook